packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts

```diff
@@ -144,6 +145,54 @@ export class ActorExtractLinksSolidTypeIndex extends ActorExtractLinks {
     return typeLinks;
   }
 
+  /**
+   * To fetch the domain of the predicate.
+   * @param predicateSubjects A dictionary of predicate and its subjects from the query.
+   * @param typeSubjects A dictionary of class type and its subjects from the query.
+   */
+  public async linkPredicateDomain(predicateSubjects: Record<string, RDF.Term>,
+    typeSubjects: Record<string, RDF.Term[]>): Promise<void> {
+    if (Object.keys(predicateSubjects).length > 0) {
+      const predicateDomains = await this.fetchPredicateDomains(Object.keys(predicateSubjects));
+
+      for (const [ predicate, subject ] of Object.entries(predicateSubjects)) {
+        const typeName = predicateDomains[predicate];
+        if (typeName) {
+          if (!typeSubjects[typeName]) {
+            typeSubjects[typeName] = [];
+          }
+          typeSubjects[typeName].push(subject);
+        }
+      }
+    }
+  }
+
+  /**
+   * To fetch the rdf type from the vocabulary if the type is not already present.
+   * @param predicateValue Array of predicate values from the query.
+   * @return predicateTypeLinks A record mapping predicate URIs to it's domain.
+   */
+  public async fetchPredicateDomains(predicateValue: string[]): Promise<Record<string, string>> {
+    const vocabSource: [IDataSource, ...IDataSource[]] =
+    <[IDataSource, ...IDataSource[]]> predicateValue;
+
+    const bindings = await this.queryEngine.queryBindings(`
+        PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+        SELECT * WHERE {
+          ?s rdfs:domain ?domain.
+        }`, {
+      sources: vocabSource,
+      [KeysRdfResolveHypermediaLinks.traverse.name]: false,
+    });
+
+    const bindingsArray = await bindings.toArray();
+    const predicateTypeLinks: Record<string, string> = {};
+    for (const binding of bindingsArray) {
+      predicateTypeLinks[binding?.get('s')?.value || ''] = binding?.get('domain')?.value || '';
```

Comment:
The `?`'s here should not be required, since the query doesn't use OPTIONALs, and will therefore guarantee bindings and values for these variables.

However, I suspect that this was added because otherwise it crashes the unit test.
This is probably because you defined the query engine in such a way that it always gives the same result for the different queries that are being executed.
Perhaps you can modify the unit tests slightly so that it checks which query is being executed (can be done in a hacky manner, by for example just checking for the presence of a certain variable), and returns different results depending on the executed query.